core/src/main/java/com/expedia/www/haystack/client/Span.java

```diff
@@ -0,0 +1,241 @@
+package com.expedia.www.haystack.client;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.atomic.AtomicReference;
+
+public class Span implements io.opentracing.Span {
+    private final Tracer tracer;
+    private final AtomicReference<SpanContext> context = new AtomicReference<>();
+    private final AtomicReference<String> operationName = new AtomicReference<>();
+    private final Map<String, Object> tags;
+    private final List<LogData> logs;
+    private final List<Reference> references;
+    private final AtomicReference<Long> duration = new AtomicReference<>();
+    private final AtomicReference<Long> endTime = new AtomicReference<>();
+    private final AtomicReference<Long> startTime = new AtomicReference<>();
+
+    private final AtomicReference<Boolean> finished = new AtomicReference<>(false);
+    private final List<RuntimeException> errors;
+    
+    public Span(Tracer tracer, String operationName, SpanContext context, long startTime, Map<String, Object> tags, List<Reference> references) {
+        this.tracer = tracer;
+        this.operationName.compareAndSet(null, operationName);
+        this.context.compareAndSet(null, context);
+        this.startTime.compareAndSet(null, startTime);
+        this.tags = new HashMap<>();
+
+        if (references == null) {
+            this.references = Collections.emptyList();
+        } else {
+            this.references = Collections.unmodifiableList(references);
+        }
+
+        this.logs = new ArrayList<>();
+
+        this.errors = new ArrayList<>();
+
+        for (Map.Entry<String, Object> entry : tags.entrySet()) {
+            this.tags.put(entry.getKey(), entry.getValue());
+        }
+    }
+
+
+	/**
+	 * Helper to record illegal access to span internals after <code>finish()</code>
+	 * has been called.
+	 *
+	 * @param format The string format to include in the execption message
+	 * @param args Any arguments needed to populate the supplied format
+	 *
+	 */
+    protected synchronized void finishedCheck(String format, Object... args) {
+        if (finished.get()) {
+            RuntimeException e = new IllegalStateException(String.format(format, args));
+            errors.add(e);
+            throw e;
+        }
+    }
+
+	/**
```

Comment:
Tabs mixed with spaces? I have IntelliJ configured to turn tabs into spaces; makes it easier to read regardless of the reader software.

Reply:
Blame eclipse for being stupid... I'll clean it up.